### DISCUSSION TITLE
Show a single responsive image

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#2] [How Might We]
[Definition of responsive image](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)

Given the same sources as #1:

An image service
A Canvas with a single image service filling the canvas
The 0.0..01% case - a canvas with partially targeted, or multiple images

...I want the component to produce the appropriate responsive image as interpreted by the browser.

**Comment by stephenwf** [comment]
[Here](https://6089966b3d7b1d0008113d68--atlas-viewer-storybook.netlify.app/iframe.html?id=static-renderer--static-render-hook-background-image&viewMode=story) is a related experiment for responsive background images. Uses a canvas internally and exports an image. However it will only load tiles required. 

There is a component in Atlas:
```
<composite-resource>
  <img src="..." height="500" width="500" />
  <img src="..." height="1000" width="1000" />
  <img src="..." height="2000" width="2000" />
</composite-resource>
```
Which treats each of the images as the same image. It will conditionally render the best size (based on the zoom level). In a situation where you _don't_ have pan/zoom, this would have the same effect as a responsive image.

However - we could/should handle these composite images with the native picture element and srcset. Where the above Atlas code, generates something like this instead:
```
<picture>
    <source srcset="..." media="(min-width: 500px)">
    <source srcset="..." media="(min-width: 1000px)">
    <source srcset="..." media="(min-width: 2000px)">
    <img src="..." alt="" />
</picture>
```
and leaves the browser to choose. 

Worth noting the limitations of the picture element. The `*-width` queries are only for the screen width, so this would become more complex in some cases. For example, if at 800px the image goes from 50% width to 100% on the screen - we would need this information before generating the `<source />` list to dial in the sizes (just as you would if you were doing it by hand).

**Comment by stephenwf** [comment]
I think related to this is an SVG renderer for Atlas. That would produce an actual image that could be saved as a `.svg` and used in place of normal images. The SVG may be made up of image tiles, or thumbnails. Those could be generated at responsive sizes and potentially be used for thumbnails. 

Exploration: Does UV/Mirador support inline svg (e.g. `data:image/svg+xml;utf8,<svg> ... </svg>`) in place of thumbnail URLs and are they displayed correctly? Could this be used to display thumbnails for composite canvases (e.g. the way TUDelft make canvases)

**Comment by stephenwf** [comment]
For the top-level API it should have different rendering "modes" so if you do not intend interacting with a canvas, there's no point in rendering a power-hungry WebGL renderer. Instead you may want a static image with some click events. 

Applying this to what a web component might look like:
```
<canvas-panel render="zoom" ... /> 
<canvas-panel render="static" ... /> 
<canvas-panel render="responsive" ... /> 
```

Which could be changed, for example on user interaction:

```
// On click... 
document.getElementById('canvas').setAttribute('render', 'zoom');
```

**Comment by tomcrane** [comment]
For responsive it needs additional attributes or child elements(?) to specify the preferred sizes. The component can also make use of the `sizes` on the service for guidance.

What about a responsive image that shows a region of the image for different widths?
E.g., the content-creator has a specific crop in mind for particular size-breaks:

```
<picture>
    <source media="(min-width: 1600px)" srcset="https://dlcs.io/.../120,850,2100,2000/1600,/0/default.jpg">
    <source media="(min-width: 700px)"  srcset="https://dlcs.io/.../250,850,1950,2000/1000,/0/default.jpg">
    <source media="(min-width: 400px)"  srcset="https://dlcs.io/.../250,850,900,1000/500,/0/default.jpg">
    <img alt="example image"               src="https://dlcs.io/.../250,850,900,1000/320,/0/default.jpg">
</picture>
```

https://tomcrane.github.io/iiif-img-tag/

There's probably no real benefit in supporting this as parameterisation of canvas panel. It would be just as much hassle to do it directly and produce the picture element yourself, on the client or server - you'd end up with the same amount of markup.

Just use responsive CP for simple cases where it is actually shorthand.

**Reply by stephenwf** [reply]
We could have render-specific options, like total sizes, max size and min size and then choose equally spaced media queries. Difficult to construct picture elements though without knowing the width relative to the page. If a picture is in a div that's 50% of the page:
```html
<source media="(min-width: 1600px)" srcset="https://dlcs.io/.../120,850,2100,2000/1600,/0/default.jpg">
```
Will be no use, instead something like:
```html
<source media="(min-width: 1600px)" srcset="https://dlcs.io/.../120,850,2100,2000/800,/0/default.jpg">
```
With 800 width.

So we could either calculate that, or let the user choose a "ratio" or similar. 

I think it might complicate the user-facing API a little bit, but could just require some design.

**Comment by tomcrane** [comment]
Proposal

Extends the `img-service` and `canvas-panel` tags from https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/1#discussioncomment-1157958

The purpose of this tag is to get rid of a lot of the fiddling about with `srcset` attributes that you need to do when using the HTML5 [picture element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture). The tag knows about available sizes because it has the image service (or the canvas). There is typically a media query involved, too:

`media="(min-width: 600px)"`

So what info would we still want the user to provide?

```html

// use the service sizes property, if present, and/or default "break points" to generate a sensible picture element without any further thinking for the developer:
<img-service src="http://..." render="responsive" />

// most of the time, it should be OK just to use the above form and let the tag do the hard work.

// provide media queries and corresponding size values; the syntax is the size slot in the Image API.
<img-service src="https://..." render="responsive">
  <source media="(min-width: 800px)" size="800,">
  <source media="(min-width: 1600px)" size="1600,">
</img-service>

// the media queries just pass through to the rendered <picture> element, we're not trying to be clever and interpret them ourselves.

// you can also specify regions (region defaults to full):

<img-service src="http://..." render="responsive">
  <source media="(min-width: 800px)" region="120,850,2100,2000", size="800,">
  <source media="(min-width: 1600px)" region="full" size="1600,">
</img-service>

// note that while the rendered picture element would include an <img /> tag, we onbiously don't need one in this component, it has all the info it needs.

// What about when this is a canvas?
// can we use the same syntax completely even though we're not necessarily dealing with an image service? Region and size work for 2D canvases too..

<canvas-panel iiif-content="..">
  <source media="(min-width: 800px)" region="120,850,2100,2000", size="800,">
  <source media="(min-width: 1600px)" region="full" size="1600,">
</canvas-panel>
```
